File: ghcopilotauth.py
import time
import requests

CLIENT_ID = "Iv1.b507a08c87ecfe98"
DEVICE_CODE_URL = "https://github.com/login/device/code"
ACCESS_TOKEN_URL = "https://github.com/login/oauth/access_token"
COPILOT_API_KEY_URL = "https://api.github.com/copilot_internal/v2/token"
HEADERS = {
    "Accept": "application/json",
    "User-Agent": "GitHubCopilotChat/0.26.7",
}

def get_device_code():
    print("Requesting device code from GitHub...")
    response = requests.post(
        DEVICE_CODE_URL,
        headers=HEADERS,
        json={
            "client_id": CLIENT_ID,
            "scope": "read:user"
        }
    )
    response.raise_for_status()
    return response.json()

def poll_for_access_token(device_code, interval, expires_in):
    print("Waiting for user to authorize...")
    for _ in range(int(expires_in / interval)):
        time.sleep(interval)
        response = requests.post(
            ACCESS_TOKEN_URL,
            headers=HEADERS,
            json={
                "client_id": CLIENT_ID,
                "device_code": device_code,
                "grant_type": "urn:ietf:params:oauth:grant-type:device_code",
            },
        )
        if response.status_code != 200:
            continue

        data = response.json()
        if "access_token" in data:
            return data["access_token"]
        elif data.get("error") == "authorization_pending":
            continue
        else:
            raise Exception(f"Authorization failed: {data}")
    raise TimeoutError("Authorization timed out.")

def get_copilot_token(access_token):
    print("Retrieving Copilot API token...")
    headers = {
        **HEADERS,
        "Authorization": f"Bearer {access_token}",
        "Editor-Version": "vscode/1.99.3",
        "Editor-Plugin-Version": "copilot-chat/0.26.7",
    }
    response = requests.get(COPILOT_API_KEY_URL, headers=headers)
    response.raise_for_status()
    data = response.json()
    return data["token"]

def main():
    device_info = get_device_code()
    print(f"\n👉 Visit {device_info['verification_uri']} and enter code: {device_info['user_code']}\n")

    access_token = poll_for_access_token(
        device_info["device_code"],
        device_info.get("interval", 5),
        device_info.get("expires_in", 900),
    )

    copilot_token = get_copilot_token(access_token)
    print("\n✅ GitHub Copilot API Key:")
    print(copilot_token)

if __name__ == "__main__":
    main()

File: ghcopilotauth.js
const axios = require('axios');

const CLIENT_ID = "Iv1.b507a08c87ecfe98";
const DEVICE_CODE_URL = "https://github.com/login/device/code";
const ACCESS_TOKEN_URL = "https://github.com/login/oauth/access_token";
const COPILOT_API_KEY_URL = "https://api.github.com/copilot_internal/v2/token";

const HEADERS = {
  Accept: "application/json",
  "User-Agent": "GitHubCopilotChat/0.26.7"
};

async function getDeviceCode() {
  console.log("Requesting device code from GitHub...");
  const res = await axios.post(DEVICE_CODE_URL, {
    client_id: CLIENT_ID,
    scope: "read:user"
  }, { headers: HEADERS });
  return res.data;
}

async function pollForAccessToken(device_code, interval, expires_in) {
  console.log("Waiting for user to authorize...");
  const totalTries = Math.floor(expires_in / interval);
  
  for (let i = 0; i < totalTries; i++) {
    await new Promise(resolve => setTimeout(resolve, interval * 1000));

    try {
      const res = await axios.post(ACCESS_TOKEN_URL, {
        client_id: CLIENT_ID,
        device_code: device_code,
        grant_type: "urn:ietf:params:oauth:grant-type:device_code"
      }, { headers: HEADERS });

      if (res.data.access_token) {
        return res.data.access_token;
      } else if (res.data.error === "authorization_pending") {
        continue;
      } else {
        throw new Error(`Authorization failed: ${JSON.stringify(res.data)}`);
      }
    } catch (err) {
      console.warn("Polling error:", err.message);
    }
  }
  throw new Error("Authorization timed out.");
}

async function getCopilotToken(access_token) {
  console.log("Retrieving Copilot API token...");
  const headers = {
    ...HEADERS,
    Authorization: `Bearer ${access_token}`,
    "Editor-Version": "vscode/1.99.3",
    "Editor-Plugin-Version": "copilot-chat/0.26.7"
  };

  const res = await axios.get(COPILOT_API_KEY_URL, { headers });
  return res.data.token;
}

async function main() {
  try {
    const deviceInfo = await getDeviceCode();
    console.log(`\n👉 Visit ${deviceInfo.verification_uri} and enter code: ${deviceInfo.user_code}\n`);

    const accessToken = await pollForAccessToken(
      deviceInfo.device_code,
      deviceInfo.interval || 5,
      deviceInfo.expires_in || 900
    );

    const copilotToken = await getCopilotToken(accessToken);
    console.log("\n✅ GitHub Copilot API Key:");
    console.log(copilotToken);
  } catch (err) {
    console.error("❌ Error:", err.message);
  }
}

main();
